Libraries/LibWasm/AbstractMachine/BytecodeInterpreter.cpp

```diff
@@ -391,1286 +2515,359 @@ double BytecodeInterpreter::read_value<double>(ReadonlyBytes data)
+                    if (nop_ip < ip.value())
+                        ++offset;
+                    else
+                        break;
+                }
+                return offset;
+            };
+
+            InstructionPointer end_ip = ptr->end_ip.value() - offset_accumulated - offset_to(ptr->end_ip - ptr->else_ip.has_value());
+            auto else_ip = ptr->else_ip.map([&](InstructionPointer const& ip) -> InstructionPointer { return ip.value() - offset_accumulated - offset_to(ip - 1); });
+            auto instruction = *result.dispatches[i].instruction;
+            instruction.arguments() = Instruction::StructuredInstructionArgs {
+                .block_type = ptr->block_type,
+                .end_ip = end_ip,
+                .else_ip = else_ip,
+            };
+            result.extra_instruction_storage.append(move(instruction));
+            result.dispatches[i].instruction = &result.extra_instruction_storage.unsafe_last();
         }
-        configuration.label_stack().append(Label(arity, ip.value() + 1, configuration.value_stack().size() - arity));
-        return;
     }
-    case Instructions::if_.value(): {
-        size_t arity = 0;
-        size_t param_arity = 0;
-        auto& args = instruction.arguments().get<Instruction::StructuredInstructionArgs>();
-        switch (args.block_type.kind()) {
-        case BlockType::Empty:
-            break;
-        case BlockType::Type:
-            arity = 1;
-            break;
-        case BlockType::Index: {
-            auto& type = configuration.frame().module().types()[args.block_type.type_index().value()];
-            arity = type.results().size();
-            param_arity = type.parameters().size();
-        }
-        }
+    for (auto index : nops_to_remove_copy)
+        result.dispatches.remove(index);
+
+    struct Value {
+        size_t id;
+        size_t definition_index;
+        Vector<size_t> uses;
+        size_t last_use = 0;
+    };
+
+    struct ActiveReg {
+        size_t value_id;
+        size_t end;
+        Dispatch::RegisterOrStack reg;
+    };
 
-        auto value = configuration.value_stack().take_last().to<i32>();
-        auto end_label = Label(arity, args.end_ip.value(), configuration.value_stack().size() - param_arity);
-        if (value == 0) {
-            if (args.else_ip.has_value()) {
-                configuration.ip() = args.else_ip.value();
-                configuration.label_stack().append(end_label);
-            } else {
-                configuration.ip() = args.end_ip.value() + 1;
+    HashMap<size_t, Value> values;
+    Vector<size_t> value_stack;
+    size_t next_value_id = 0;
+    HashMap<size_t, size_t> instr_to_output_value;
+    HashMap<size_t, Vector<size_t>> instr_to_input_values;
+    HashMap<size_t, Vector<size_t>> instr_to_dependent_values;
+
+    Vector<size_t> forced_stack_values;
+
+    Vector<size_t> parent;
+    Vector<size_t> rank;
+    Vector<size_t> final_roots;
```

Comment:
Maybe annotate what's a ValueID/stack index and whats an IP
(like with some `typedefs` or even `DistinctNumeric`)